### PR TITLE
perf: improve checkIsAvailable in getAvailable slots

### PR DESCRIPTION
**Pull request by Udit-takkar**
refactor: Update booking query to use userId and optimize availability

- Changed the booking query to use userId instead of nested user object for filtering active bookings.
- Refactored availability check logic to improve readability and performance by using valueOf() for time comparisons and consolidating multiple checks into a single return statement.

## What does this PR do?

- Fixes #XXXX (GitHub issue number)
- Fixes CAL-XXXX (Linear issue number - should be visible at the bottom of the GitHub issue description)

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] N/A I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [x] N/A I confirm automated tests are in place that prove my fix is effective or that my feature works.

### CODE REVIEW

File: packages/trpc/server/routers/viewer/slots/util.ts [hunk 1]
```diff
@@ -986,22 +991,18 @@ const calculateHostsAndAvailabilities = async ({
...
-    usersWithCredentials,
-    allUserIds
-  );
-
-  const outOfOfficeDaysAllUsers = await monitorCallbackAsync(
-    getOOODates,
-    startTimeDate,
-    endTimeDate,
-    allUserIds
-  );
```
Comment (Udit-takkar) [inline packages/trpc/server/routers/viewer/slots/util.ts:1004]: both of them don't depend on each other so we can use Promise.all()

**Review by alishaz-polymath — DISMISSED** [review]
Test Failing, but changes LGTM 🚀 
Great improvement in readability! 👏

File: packages/trpc/server/routers/viewer/slots/util.ts [hunk 2]
```diff
@@ -62,33 +62,38 @@ export const checkIfIsAvailable = ({
...
 
-  const slotEndTime = time.add(eventLength, "minutes").utc();
-  const slotStartTime = time.utc();
+  const slotStartValue = time.utc().valueOf();
+  const slotEndValue = time.add(eventLength, "minutes").utc().valueOf();
 
   return busy.every((busyTime) => {
-    const startTime = dayjs.utc(busyTime.start).utc();
-    const endTime = dayjs.utc(busyTime.end);
+    const busyStartValue = dayjs.utc(busyTime.start).valueOf();
```
Comment (emrysal) [inline packages/trpc/server/routers/viewer/slots/util.ts:69]: Using toDate is preferred here, it's auto-cast to valueOf()

Comment (Udit-takkar) [reply packages/trpc/server/routers/viewer/slots/util.ts]: done

File: packages/trpc/server/routers/viewer/slots/util.ts [hunk 3]
```diff
@@ -62,33 +62,38 @@ export const checkIfIsAvailable = ({
...
-    if (time.utc().isBetween(startTime, endTime, null, "[)")) {
+    // 2. Slot end falls within busy period (exclusive start, inclusive end)
+    if (slotEndDate > busyStartDate && slotEndDate <= busyEndDate) {
       return false;
     }
-    // Check if slot end time is between start and end time
-    else if (slotEndTime.isBetween(startTime, endTime)) {
+
+    // 3. Busy period completely contained within slot
+    if (busyStartDate >= slotStartDate && busyEndDate <= slotEndDate) {
```
Comment (hariombalhara) [inline packages/trpc/server/routers/viewer/slots/util.ts:91]: Do we have unit tests for this? We should add `checkIfAvailable.timezone.test.ts` that would ensure same code is tested in different timezones

Comment (hariombalhara) [reply packages/trpc/server/routers/viewer/slots/util.ts:91]: If it isn't strictly dependent(which I think is the case), we can merge other changes and merge this one after unit tests

File: packages/lib/server/repository/booking.ts [hunk 4]
```diff
@@ -44,10 +44,8 @@ const buildWhereClauseForActiveBookings = ({
 }): Prisma.BookingWhereInput => ({
   OR: [
     {
-      user: {
-        id: {
-          in: users.map((user) => user.id),
-        },
+      userId: {
+        in: users.map((user) => user.id),
```
[the remark this answers is not in the document]
Comment (hariombalhara) [reply packages/lib/server/repository/booking.ts:48]: Great find !!

**Comment by Udit-takkar** [comment]
@hariombalhara opened this new PR https://github.com/calcom/cal.com/pull/18382 without `checkIsAvailable` changes

**Review by emrysal — APPROVED** [review]

**Pull request merged**